#! /bin/sh /usr/share/dpatch/dpatch-run
## 13_add_NoFollow_feature.dpatch by Jose Carlos Medeiros <debian@psabs.com.br>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Add "Nofollow" option.

################################################
# use rel=nofollow for referrers external links
# http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=293794
# Robert Cheramy <robert@cheramy.net>
#
# OBS:
# This problem was solved in "01_basic_patch.dpatch" patch.
# But I added this patch to make possible you put <A> tags
# with "protocol://" if you wish.
# By default webalizer follow first patch that strip protocol://
# strings.
# ########
#
# Somebody is using by referrer logs for search engine spamming. They
# set the referrer field to their webpage, and then hit my server enough
# times that they end up in the top 30 referrers displayed in the
# statistics.
#
# If you look at
# http://www.kollegiegaarden.dk/webalizer/usage_200501.html#TOPREFS you
# can see that there are 13 entries with exactly 50 referrer hits, for
# pages like http://paris-hotel-las-vegas.go.ro/ and
# http://reno-hotels.go.ro/. Actually I think only 1 out of 30 items on
# that list is genuine.
#
# It would be a nice simple solution if webalizer supported rel=nofollow
# (http://www.google.com/googleblog/2005/01/preventing-comment-spam.html). If
# this was implemented then the motivation for this spam would quickly
# disappear. Since the links has no real value for calculating page
# value in the first place there seems to be no negatives involved with getting
# them disregarded by search engines.


@DPATCH@
Index: webalizer-2.01.10/README
===================================================================
--- webalizer-2.01.10.orig/README	2006-10-28 06:01:29.000000000 +0400
+++ webalizer-2.01.10/README	2006-10-28 06:07:02.000000000 +0400
@@ -870,6 +870,10 @@
               a value of zero ('0') to disable.  See the DNS.README
               file for additional information.

+NoFollow      Specifies if the attribute 'rel="nofollow"' should be added
+              to the links to the referrers. Use a value of zero ('0') to
+              strip protocol://, use ('1') to put rel=nofollow or ('2') to
+              put original string (disable this option). Default is ('1').

 Top Table Keywords
 ------------------
Index: webalizer-2.01.10/output.c
===================================================================
--- webalizer-2.01.10.orig/output.c	2006-10-28 06:01:29.000000000 +0400
+++ webalizer-2.01.10/output.c	2006-10-28 06:07:02.000000000 +0400
@@ -1389,19 +1389,35 @@
          }
          else
          {
-	    /* do not print as anchor tags to avoid referrer spamming */
-	    /* skip over any initial protocol:// */
-	    char *p = rptr->string;
-	    while (*p && isalpha(*p)) {
-	    	++p;
-	    }
-	    if (*p && strncmp(p, "://", 3) == 0) {
-		fprintf(out_fp,"%s", p+3);
-	    } else if (*p && strncmp(p, ":/", 2) == 0) {
-		fprintf(out_fp,"%s", p+1);
-	    } else {
-		fprintf(out_fp,"%s", rptr->string);
-	    }
+	    /* Added option to put rel=nofollow or just strip protocol:// */
+            if (rptr->string[0] != '-')
+            {
+               if (nofollow == 0) /* default option, strip protocol:// */
+               {
+                  /* do not print as anchor tags to avoid referrer spamming */
+                  /* skip over any initial protocol:// */
+                  char *p = rptr->string;
+                  while (*p && isalpha(*p)) {
+                        ++p;
+                  }
+                  if (*p && strncmp(p, "://", 3) == 0) {
+                     fprintf(out_fp,"%s", p+3);
+                  } else if (*p && strncmp(p, ":/", 2) == 0) {
+                     fprintf(out_fp,"%s", p+1);
+                  } else {
+                     fprintf(out_fp,"%s", rptr->string);
+	          }
+               } else if (nofollow == 1) { /* put with rel=nofollow option */
+                  fprintf(out_fp,"<A HREF=\"%s\" rel=\"nofollow\" target=\"_blank\">%s</A>",
+                          rptr->string, rptr->string);
+               } else { /* put as original ( with protocol:// ) ( I dont know why, but you can do it ) */
+                  fprintf(out_fp,"<A HREF=\"%s\">%s</A>",
+                          rptr->string, rptr->string);
+               }
+            } else {
+               fprintf(out_fp,"%s", rptr->string);
+            }
+
          }
          fprintf(out_fp,"</FONT></TD></TR>\n");
          tot_num--;
Index: webalizer-2.01.10/webalizer.c
===================================================================
--- webalizer-2.01.10.orig/webalizer.c	2006-10-28 06:01:29.000000000 +0400
+++ webalizer-2.01.10/webalizer.c	2006-10-28 06:07:02.000000000 +0400
@@ -145,6 +145,7 @@
 char    *blank_str   = "";                    /* blank string             */
 char    *dns_cache   = NULL;                  /* DNS cache file name      */
 int     dns_children = 0;                     /* DNS children (0=don't do)*/
+int     nofollow     = 1;                     /* Referrer Following (2=no)*/

 #ifdef USE_GEOIP
 int     use_geoip    = 1;                     /* Use GeoIP library        */
@@ -1542,6 +1543,7 @@
                      "GeoIP",             /* Use GeoIP library (0=no)   88  */
                      "GeoIPDatabase",     /* GeoIP database             89  */
 #endif	/* USE_GEOIP */
+                     "NoFollow"           /* Referrer Following (2=no)  90 */
                    };

    FILE *fp;
@@ -1689,6 +1691,7 @@
         case 88: use_geoip=(value[0]=='n')?0:1; break;    /* GeoIP          */
         case 89: geoip_dbase=save_opt(value); break;      /* GeoIPDatabase  */
 #endif	/* USE_GEOIP */
+        case 90: nofollow== atoi(value); break;           /* NoFollow       */
       }
    }
    fclose(fp);
Index: webalizer-2.01.10/webalizer.h
===================================================================
--- webalizer-2.01.10.orig/webalizer.h	2006-10-28 06:01:29.000000000 +0400
+++ webalizer-2.01.10/webalizer.h	2006-10-28 06:07:02.000000000 +0400
@@ -194,6 +194,7 @@
 extern char    *blank_str   ;                 /* blank string             */
 extern char    *dns_cache   ;                 /* DNS cache file name      */
 extern int     dns_children ;                 /* # of DNS children        */
+extern int     nofollow     ;                 /* Referrer Following (2=no)*/

 #ifdef USE_GEOIP
 extern int     use_geoip    ;                 /* Use GeoIP library        */
